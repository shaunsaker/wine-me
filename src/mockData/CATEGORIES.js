const CATEGORIES = {
  wine: {
    order: 1,
    validation: {
      type: 'number',
      min: 1,
      max: 5,
    },
    name: 'Wine',
    question_text: 'How would you rate the wine?',
    is_rating: true,
  },
  child_friendly: {
    order: 2,
    validation: {
      type: 'string',
    },
    name: 'Child friendly',
    question_text: 'Was it child friendly?',
    options: ['Yes', 'No'],
  },
  price: {
    order: 3,
    validation: {
      type: 'number',
      min: 0,
    },
    name: 'Price',
    question_text: 'What did the wine tasting cost?',
  },
};

export default CATEGORIES;
