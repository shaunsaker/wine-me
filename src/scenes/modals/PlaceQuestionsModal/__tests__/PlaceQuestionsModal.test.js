import React from 'react';
import renderer from 'react-test-renderer';

import { PlaceQuestionsModal } from '..';
import PLACE from '../../../../mockData/PLACE';
import CATEGORIES from '../../../../mockData/CATEGORIES';

jest.mock('../../../../components/PlaceQuestionsList', () => 'PlaceQuestionsList');

// Fixes _bezier is not a function bug
jest.useFakeTimers();

describe('PlaceQuestionsModal', () => {
  const spies = [];
  const dispatch = jest.fn();
  const categoryID = 'wine';
  const rules = CATEGORIES[categoryID].validation;
  const validValue = rules.max - 1;
  const nextSlideIndex = 1;
  const systemMessage = 'Thanks for the feedback.';

  describe('renders', () => {
    it('renders with minimum required props', () => {
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );

      expect(component).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    describe('should handle onSetValue', () => {
      it('with a valid value', () => {
        spies[0] = jest.spyOn(PlaceQuestionsModal.prototype, 'setValue');
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();

        instance.onSetValue(categoryID, validValue);

        expect(spies[0]).toHaveBeenCalledWith(categoryID, validValue);
      });

      it('with an invalid value', () => {
        spies[0] = jest.spyOn(PlaceQuestionsModal.prototype, 'setValue');
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();

        instance.onSetValue(categoryID, rules.max + 1);

        expect(spies[0]).not.toHaveBeenCalled();
      });
    });

    describe('should handle validateValue', () => {
      it('when value is NaN', () => {
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();
        const isValid = instance.validateValue(NaN, rules);

        expect(isValid).toEqual(false);
      });

      it('when the value is not of number type', () => {
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();
        const isValid = instance.validateValue('string', rules);

        expect(isValid).toEqual(false);
      });

      it('when the value lower than min', () => {
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();
        const isValid = instance.validateValue(rules.min - 1, rules);

        expect(isValid).toEqual(false);
      });

      it('when the value is higher than max', () => {
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();
        const isValid = instance.validateValue(rules.max + 1, rules);

        expect(isValid).toEqual(false);
      });
    });

    it('should handle setValue', () => {
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const instance = component.getInstance();
      const categoryName = CATEGORIES[categoryID].name;

      instance.setValue(categoryID, validValue);

      expect(instance.state[categoryName]).toEqual(validValue);
    });

    describe('should handle onSubmit', () => {
      it('when it is the last slide', () => {
        spies[0] = jest.spyOn(PlaceQuestionsModal.prototype, 'dismissKeyboard');
        spies[1] = jest.spyOn(PlaceQuestionsModal.prototype, 'saveUserFeedback');
        spies[2] = jest.spyOn(PlaceQuestionsModal.prototype, 'setSystemMessage');
        spies[3] = jest.spyOn(PlaceQuestionsModal.prototype, 'navigate');
        spies[4] = jest.spyOn(PlaceQuestionsModal.prototype, 'setSlideIndex');
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();

        // SETUP
        const numberOfCategories = Object.keys(CATEGORIES).length;
        const lastSlide = numberOfCategories - 1;
        instance.setSlideIndex(lastSlide);
        spies[4].mockClear();

        instance.onSubmit();

        expect(spies[0]).toHaveBeenCalled();
        expect(spies[1]).toHaveBeenCalled();
        expect(spies[2]).toHaveBeenCalledWith(systemMessage);
        expect(spies[3]).toHaveBeenCalled();
        expect(spies[4]).not.toHaveBeenCalled();
      });

      it('when it is not the last slide', () => {
        spies[0] = jest.spyOn(PlaceQuestionsModal.prototype, 'dismissKeyboard');
        spies[1] = jest.spyOn(PlaceQuestionsModal.prototype, 'saveUserFeedback');
        spies[2] = jest.spyOn(PlaceQuestionsModal.prototype, 'setSystemMessage');
        spies[3] = jest.spyOn(PlaceQuestionsModal.prototype, 'navigate');
        spies[4] = jest.spyOn(PlaceQuestionsModal.prototype, 'setSlideIndex');
        const component = renderer.create(
          <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
        );
        const instance = component.getInstance();

        instance.onSubmit();

        expect(spies[0]).toHaveBeenCalled();
        expect(spies[1]).toHaveBeenCalled();
        expect(spies[2]).not.toHaveBeenCalled();
        expect(spies[3]).not.toHaveBeenCalled();
        expect(spies[4]).toHaveBeenCalledWith(nextSlideIndex);
      });
    });

    it('should handle onSkip', () => {
      spies[0] = jest.spyOn(PlaceQuestionsModal.prototype, 'setSlideIndex');
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const instance = component.getInstance();

      instance.onSkip();

      expect(spies[0]).toHaveBeenCalledWith(nextSlideIndex);
    });

    it('should handle saveUserFeedback', () => {
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const instance = component.getInstance();

      // SETUP
      instance.setValue(categoryID, validValue);

      instance.saveUserFeedback();

      expect(dispatch).toHaveBeenCalled();
      expect(dispatch).toMatchSnapshot();
    });

    it('should handle dismissKeyboard', () => {
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const instance = component.getInstance();

      instance.dismissKeyboard();
    });

    it('should handle setSlideIndex', () => {
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const instance = component.getInstance();

      instance.setSlideIndex(nextSlideIndex);

      expect(instance.state.slideIndex).toEqual(nextSlideIndex);
    });

    it('should handle setSystemMessage', () => {
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const instance = component.getInstance();

      instance.setSystemMessage(systemMessage);

      expect(dispatch).toHaveBeenCalled();
      expect(dispatch).toMatchSnapshot();
    });

    it('should handle navigate', () => {
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const instance = component.getInstance();

      instance.navigate();
    });
  });

  describe('actions', () => {
    it('should call onSkip on Link button press', () => {
      spies[0] = jest.spyOn(PlaceQuestionsModal.prototype, 'onSkip');
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const { root } = component;
      const targetComponent = root.findByProps({ testID: 'placeQuestionsModal.button.skip' });

      targetComponent.props.handlePress();

      expect(spies[0]).toHaveBeenCalled();
    });

    it('should call onSubmit on Submit button press', () => {
      spies[0] = jest.spyOn(PlaceQuestionsModal.prototype, 'onSubmit');
      const component = renderer.create(
        <PlaceQuestionsModal place={PLACE} categories={CATEGORIES} dispatch={dispatch} />,
      );
      const { root } = component;
      const targetComponent = root.findByProps({ testID: 'placeQuestionsModal.button.submit' });

      targetComponent.props.handlePress();

      expect(spies[0]).toHaveBeenCalled();
    });
  });

  afterEach(() => {
    spies.forEach((spy) => {
      if (spy) {
        spy.mockClear();
      }
    });
    dispatch.mockClear();
  });
});
