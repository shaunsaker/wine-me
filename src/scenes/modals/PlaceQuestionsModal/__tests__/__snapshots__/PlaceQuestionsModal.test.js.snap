// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceQuestionsModal methods should handle saveUserFeedback 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "meta": Object {
          "pathParts": Array [
            "user_feedback",
          ],
        },
        "payload": Object {
          "document": Object {
            "category_name": "Wine",
            "date": 1533309837130,
            "place_id": "1",
            "uid": undefined,
            "value": 4,
          },
        },
        "type": "addDocument",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`PlaceQuestionsModal methods should handle setSystemMessage 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Object {
          "message": "Thanks for the feedback.",
        },
        "type": "SET_SYSTEM_MESSAGE",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`PlaceQuestionsModal renders renders with minimum required props 1`] = `
<View
  style={
    Object {
      "backgroundColor": "rgba(0, 0, 0, 0.70)",
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "transform": Array [
        Object {
          "translateY": 1334,
        },
      ],
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
        "margin": 30,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 5,
          "maxWidth": 680,
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#48466D",
            "borderColor": null,
            "borderWidth": 0,
            "elevation": 2,
            "justifyContent": "center",
            "paddingVertical": 10,
            "shadowColor": "#212121",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.17,
            "shadowRadius": 2,
          }
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          numberOfLines={1}
          style={
            Object {
              "color": "#FFFFFF",
              "fontFamily": "Nunito-SemiBold",
              "fontSize": 17,
              "marginLeft": 15,
              "marginRight": 39.5,
            }
          }
        >
          Bartinney Tasting Shed
        </Text>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 7.5,
              "position": "absolute",
              "right": 0,
            }
          }
          testID="lightbox.button.close"
        >
          <Text
            accessible={true}
            allowFontScaling={false}
            ellipsizeMode="tail"
            style={
              Array [
                Object {
                  "color": undefined,
                  "fontSize": 12,
                },
                Object {
                  "color": "#FFFFFF",
                  "fontSize": 24,
                },
                Object {
                  "fontFamily": "Material Icons",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            Óóç
          </Text>
        </View>
      </View>
      <View
        style={Object {}}
      >
        <PlaceQuestionsList
          data={
            Array [
              Object {
                "id": "wine",
                "is_rating": true,
                "name": "Wine",
                "order": 1,
                "question_text": "How would you rate the wine?",
                "validation": Object {
                  "max": 5,
                  "min": 1,
                  "type": "number",
                },
                "value": undefined,
              },
              Object {
                "id": "child_friendly",
                "name": "Child friendly",
                "options": Array [
                  "Yes",
                  "No",
                ],
                "order": 2,
                "question_text": "Was it child friendly?",
                "validation": Object {
                  "type": "string",
                },
                "value": undefined,
              },
              Object {
                "id": "price",
                "name": "Price",
                "order": 3,
                "question_text": "What did the wine tasting cost?",
                "validation": Object {
                  "min": 0,
                  "type": "number",
                },
                "value": undefined,
              },
            ]
          }
          handleSetValue={[Function]}
          slideIndex={0}
        />
        <View
          style={
            Object {
              "marginHorizontal": 15,
              "marginVertical": 10,
            }
          }
        >
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#E23E57",
                "borderColor": null,
                "borderRadius": 5,
                "borderWidth": 0,
                "elevation": 2,
                "opacity": 0.33,
                "paddingHorizontal": 15,
                "paddingVertical": 10,
                "shadowColor": "#212121",
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.17,
                "shadowRadius": 2,
              }
            }
            testID="placeQuestionsModal.button.submit"
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Array [
                  Object {
                    "color": "#212121",
                    "fontFamily": "Nunito-SemiBold",
                    "fontSize": 17,
                  },
                  Object {
                    "color": "#FFFFFF",
                  },
                  Object {},
                ]
              }
            >
              Next
            </Text>
          </View>
        </View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginHorizontal": 15,
              "marginVertical": 10,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#212121",
                "fontFamily": "Nunito-Regular",
                "fontSize": 17,
              }
            }
          >
            1 / 3
          </Text>
          <View
            accessible={true}
            isTVSelectable={true}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID="placeQuestionsModal.button.skip"
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
              style={
                Object {
                  "color": "#3D84A8",
                  "fontFamily": "Nunito-SemiBold",
                  "fontSize": 17,
                }
              }
            >
              Skip
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
